python/lsst/cp/pipe/cpFilterScan.py

```diff
@@ -0,0 +1,377 @@
+        Parameters
+        ----------
+        inputExpHandles : `list` [`lsst.daf.butler.DeferredDatasetHandle`]
+            Input list of exposure handles to combine.
+        inputElectrometerHandles : `list` [`lsst.daf.butler.DeferredDatasetHandle`], optional # noqa W505
+            Input list of electrometer/photodiode measurement handles
+            to combine.
+
+        Returns
+        -------
+        results : `lsst.pipe.base.Struct`
+            The results struct containing:
+
+            ``outputData``
+                Final combined filter scan, with a single table
+                containing the measured throughput for all input
+                filters at the various wavelength values indicated in
+                the exposure's observationReason
+                (`astropy.table.Table`).
+        """
+        filterScanResults = {}
+        filterSet = set()
+        efdClient = CpEfdClient(efdInstance=self.config.efdClientInstance)
+        monochromatorData = efdClient.getEfdMonochromatorData()
+
+        for handle in inputExpHandles:
+            visitInfo = handle.get(component="visitInfo")
+            metadata = handle.get(component="metadata")
+
+            physical_filter = handle.dataId['physical_filter']
+            exposureId = handle.dataId['exposure']
+            exposureTime = visitInfo.exposureTime
+
+            values = [v for k, v in metadata.items() if "LSST ISR FINAL MEAN" in k]
+
+            scaleFactor = exposureTime
+            if self.config.useElectrometer:
+                for eHandle in inputElectrometerHandles:
+                    if eHandle.dataId.exposure == exposureId:
+                        charge = eHandle.get().integrate()
+                        if charge > 0.0 and np.isfinite(charge):
+                            scaleFactor /= charge
+                        break
+
+            scan = {
+                'physical_filter': physical_filter,
+                'scale': scaleFactor,
+                'flux': np.median(values),
+            }
+            filterSet.add(physical_filter)
+
+            _, key = efdClient.parseMonochromatorStatus(monochromatorData,
+                                                        visitInfo.date.toString(DateTime.TAI))
+            key = float(key)
+            if key in filterScanResults:
+                filterScanResults[key].append(scan)
+            else:
+                filterScanResults[key] = [scan]
+
+        filterScan = []
+        for wavelength in filterScanResults.keys():
+            scans = filterScanResults[wavelength]
+
+            referenceScan = [x for x in scans if x['physical_filter'] == self.config.referenceFilter]
+            if len(referenceScan) == 0:
+                # No reference scan at this wavelength.
+                continue
+            referenceScan = referenceScan[0]
+            referenceValue = referenceScan['scale'] / referenceScan['flux']
+
+            wavelengthScan = {'wavelength': float(wavelength)}
```

Comment:
I'm having trouble parsing what these dicts are supposed to contain.  Below I see that it always sets the wavelengthScan to nan and then checks if it's finite and how is it ever finite?

Reply:
After the first loop, we have a `filterScanResults` dictionary, keyed by the wavelength of the monochromator setting, and containing a list of scans.  Each scan in that list maps to one input exposure, and contains the filter of that exposure, the measured exposure flux, and a scale factor to normalize exposure time and the electrometer's measurement of the illumination.
This second loop then iterates over all scans taken at a given monochromator wavelength, with the intention of pairing them up.  In the generic case, there can be any number of filters taken at that wavelength, so the set of all filters observed at that wavelength is determined, and the throughput of all found filters at that wavelength are set to NaN.  Ideally, the wavelengthScan looks like `{wavelength: 12345, g: NaN, r: NaN, i: NaN, empty: NaN}`.  The code then loops over the entries of `scans`, and if the value for a particular filter is NaN, that filter has not yet been filled, and is then set to the appropriate throughput.  If the value for that filter is not NaN, then a previous scan already populated that field, and the log message is triggered.  This task has no way to know which of the multiple scans are correct, but if there are matching electrometer values for everything, they should normalize to the same value.  There's something of a check in that `empty` is both a filter with a scan and the reference, so the throughput of the reference filter should be universally 1.0.

Reply:
This is all wonderfully clear and very helpful.  If you could put it into the code comments then in 6 months we'll have a lot easier time remembering what it is doing.

Reply:
Let me know if further comments would help.